lib/core/src/Cardano/Wallet.hs

```diff
@@ -692,34 +695,45 @@ newWalletLayer tracer bp db nw tl = do
                         then Ready
                         else Restoring progress'
 
-            let txs_cps = applyBlocks @s @t (NE.toList blocks) cp
-            let txs = fold $ fst <$> txs_cps
-
-            let cpLast = fromMaybe cp $ lastMay $ snd <$> txs_cps
-            let (Quantity bhLast) = blockHeight cpLast
+            -- NOTE:
+            -- Always store the last checkpoint from the batch and all new
+            -- transactions.
+            let cpLast = NE.last cps
+            let Quantity bhLast = blockHeight cpLast
+            let meta' = calculateMetadata (view #slotId $ currentTip cpLast)
+            DB.putCheckpoint db (PrimaryKey wid) cpLast
+            DB.putTxHistory db (PrimaryKey wid) newTxs
+            DB.putWalletMeta db (PrimaryKey wid) meta'
+
+            -- We only need to create checkpoints for blocks that are
+            -- "unstable".
+            -- We define unstable blocks as blocks that `k` blocks from the
+            -- network tip, where k is the epoch stability parameter.
+            -- NOTE:
+            -- We cast `k` and block heights to 'Integer' since at a given point
+            -- in time, `k` may be greater than the tip.
+            let (Quantity k) = epochStability
+            let (Quantity tipBh) = blockHeight cpLast
```

Comment:
I wonder if it might not turn out to be a bit tricky. `cpLast` shouldn't necessarily be the network tip, because `restoreBlocks` is called with batches of blocks.

Restore benchmarks seem slower, I believe because of above:

```
All results:
  restore testnet seq: 40.67 s
  restore testnet 10% ownership: 40.86 s
Benchmark restore: FINISH
Completed 2 action(s).
```

But I'm not on the clear with how all the pieces should fit together in the grand scheme of things.

Reply:
Man.. Sorry, I am distracted this morning. So, yes, of course the last checkpoint in a particular batch isn't the network tip >.> ... Now, yeah, this is why I started using the `nodeTip` initially but then, we only have its slot number and not its block number. 

I am not sure what is safe to assume here. In practice we won't have much empty blocks but they can happen. In theory, it can even be that 2160 blocks are empty which means that rolling back of `k` blocks would require a checkpoint very far away. 

Yet, I believe we can sort it out nicely by: 

- Making regular sparse checkpoints (as this does), every batch
- Having indeed the last `k` checkpoints for each slots near the network tip
- Upon rollback, selecting the nearest checkpoint. 

In practice, we should always end up having enough checkpoints, even if there's like only 10% of blocks because, big rollbacks are less likely than small ones. If it happens, then we'll simply rollback to the nearest checkpoint and recover from there, even if it's many blocks from the hypothetical point of intersection.

Reply:
We could also lookup the node's own chain height: 
 
https://input-output-hk.github.io/jormungandr/jcli/rest.html#node-stats

Reply:
> We could also lookup the node's own chain height:

Hm… That's interesting

For me it seems very empty though:
```
blockRecvCnt: 0
lastBlockFees: 0
lastBlockSum: 0
lastBlockTime: ~
lastBlockTx: 0
txRecvCnt: 0
uptime: 207
```
despite the tip blockId changing.

I also imagined we could check if the last blockId in the batch matches the blockId of the network tip. But then we might discover too late that the unstable blocks started in the previous batch…

Reply:
From the rust team: 

- Chain stats is currently not well updated 
- The `blockRecvCnt` does not represent the chain height, but instead, the total number of blocks received from all peers (so if 10 peers sends 10 blocks each, `blockRecvCnt` equals 100).

Reply:
Also, the binary block header format actually contains the chain length:

```
Block Header

The header is a small piece of data, containing enough informations for validation and network deduplication and a strong signed cryptographic link to the content.

Common (2 * 64 bits + 1 * 32 bits + 2 * 256 bits = 84 bytes):
    Size of Header: 2 bytes (16 bits): Maximum header is thus 64K not including the block content
    Version of block: 2 bytes (16 bits)
    Size of Content: 4 bytes (32 bits)
    Block Date: Epoch (4 bytes, 32 bits) + Slot-id (4 bytes - 32 bits)
    Chain length (number of ancestor blocks; first block has chain length 0): 4 bytes (32 bits)
    Hash of content H(Content) (32 bytes - 256 bits)
    Parent Header hash : 32 bytes (256 bits)
```

Reply:
😮 !

Reply:
so, I've extracted the `chainLength` from the tip block header in Jormungandr network layer, such that we now have:

```hs
    , networkTip
        :: ExceptT ErrNetworkTip m (BlockHeader, Quantity "block" Natural)
```

instead of just:

```hs
    , networkTip
        :: ExceptT ErrNetworkTip m BlockHeader
```

For `http-bridge`, it "defaults" to the node's flat slot which should be good-enough. Now, we can reliably use the node's chain height during restoration to figure out whether we are in an unstable zone.